README.md
qsqla
===============================

[![Build Status](https://travis-ci.org/blue-yonder/qsqla.svg?branch=master)](https://travis-ci.org/blue-yonder/qsqla)

qSQLA is a query builder for SQLAlchemy Core Selectables 

Installation / Usage
--------------------

To install use pip:

    $ pip install qsqla


Or clone the repo:

    $ git clone https://github.com/blue-yonder/qsqla.git
    $ python setup.py install

Example
-------


```python

from sqlalchemy import create_engine, MetaData, Table, Column, Integer, String


db = create_engine("sqlite:////:memory:")
md = MetaData(bind=db)

table = Table('user', 
              md,
              Column('id', Integer, primary_key=True),
              Column('name', String(16), nullable=False),
              Column('age', Integer)
              )

sel = table.select()

from qsqla.query import query, build_filters
filter =  build_filters({"id__eq":1})
stm = query(sel, filter)

```


CHANGELOG
Next Release
============

0.3.2
=====

- Handle case where `limit` is set but `upper_bound_limit` is `None`

0.3.1
=====

- Added `upper_bound_limit` for specifying (or disabling) the current default upper bound of 10000 rows.

0.3.0
=====

- Added `with` operator for n-to-m relationships. This can only be used with the ORM. 

0.2.0
=====

- Added SQLAlchemy ORM support

0.1.2
=====

- Bugfix: Fixed type conversion of typedecorated types

0.1.1
=====

- Added support for custom types which use sqlalchemy.types.TypeDecorator.
- Added support for ilike and not_ilike filter
- Use postgresql in travis tests

0.1.0
=====

- Initial Release
- Support for python 2.7, 3.5 and 3.6-dev


tests/test_qsqla.py
import unittest
import os

from datetime import date, datetime, timedelta
from operator import itemgetter
from sqlalchemy import (MetaData, Table, Column, DateTime, Integer, String,
                        ForeignKey, create_engine, types, select)
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, relationship

import qsqla.query as qsqla


class CustomDateTime(types.TypeDecorator):
    impl = DateTime


Base = declarative_base()


user_pet = Table('user_pet',
                 Base.metadata,
                 Column('user_u_id', Integer, ForeignKey('user_table.u_id')),
                 Column('pet_p_id', Integer, ForeignKey('pet.p_id'))
                 )

class Location(Base):
    __tablename__ = 'location'
    l_id = Column(Integer, primary_key=True)
    l_name = Column(String(16))
    l_date = Column(CustomDateTime)


class User(Base):
    __tablename__ = 'user_table'
    u_id = Column(Integer, primary_key=True)
    u_name = Column(String(16))
    u_l_id = Column(ForeignKey(Location.l_id))
    location = relationship(Location)
    pets = relationship("Pet", secondary=user_pet)
    u_date = Column(CustomDateTime)


class Pet(Base):
    __tablename__ = 'pet'
    p_id = Column(Integer, primary_key=True)
    p_name = Column(String(16))
    p_date = Column(CustomDateTime)


class TestSqlaSplitOperator(unittest.TestCase):
    def test_without_operator(self):
        param = "field"
        name, operator = qsqla.split_operator(param)
        self.assertEquals(name, param)
        self.assertEquals(operator, "eq")

    def test_with_operator(self):
        param = "field__eq"
        name, operator = qsqla.split_operator(param)
        self.assertEquals(name, "field")
        self.assertEquals(operator, "eq")

    def test_with_operator_fancy_name(self):
        param = "field_with__underscore__eq"
        name, operator = qsqla.split_operator(param)
        self.assertEquals(name, "field_with__underscore")
        self.assertEquals(operator, "eq")


class TestSqlaFilter(unittest.TestCase):
    def test_empty(self):
        f = {}
        filters = qsqla.build_filters(f)
        self.assertEquals(filters, [])

    def test_one_without_operator(self):
        f = {"age": 55}
        filters = qsqla.build_filters(f)
        self.assertEquals(filters, [{"name": "age", "op": "eq", "val": 55}])

    def test_one_with_operator(self):
        f = {"age__gt": 55}
        filters = qsqla.build_filters(f)
        self.assertEquals(filters, [{"name": "age", "op": "gt", "val": 55}])

    def test_multiple(self):
        f = {"age__eq": 55, "surname__like": "%joe%"}
        filters = qsqla.build_filters(f)
        expected = [{"name": "surname", "op": "like", "val": "%joe%"},
                    {"name": "age", "op": "eq", "val": 55}]

        self.assertEquals(sorted(filters, key=itemgetter("name")), 
                          sorted(expected, key=itemgetter("name")))


class DBTestCase(unittest.TestCase):
    def setUp(self):
        dsn = os.environ.get("ENV_DSN", "sqlite:///:memory:")
        self.engine = create_engine(dsn)
        self.db = self.engine.connect()
        self.session = sessionmaker()(bind=self.db)
        self.now = datetime.now()
        Base.metadata.create_all(self.db)
        self.user = User.__table__
        self.location = Location.__table__
        l1 = Location(l_name='Karlsruhe', l_date=self.now)
        l2 = Location(l_name='Stuttgart', l_date=self.now)
        self.session.add(l1)
        self.session.add(l2)
        p1 = Pet(p_name="Beethoven", p_date=self.now)
        p2 = Pet(p_name="Hooch", p_date=self.now)
        p3 = Pet(p_name="Sissy", p_date=self.now)
        self.session.add(p1)
        self.session.add(p2)
        self.session.add(p3)
        u1 = User(u_name='Micha', location=l1, u_date=self.now)
        u2 = User(u_name='Oli', location=l1, u_date=self.now)
        u3 = User(u_name='Tom', location=l2, u_date=self.now)
        u1.pets.append(p1)
        u1.pets.append(p2)
        u2.pets.append(p2)
        u3.pets.append(p3)
        self.session.add(u1)
        self.session.add(u2)
        self.session.add(u3)
        self.session.commit()
        self.joined_select = self.location.join(
            self.user, self.location.c.l_id == self.user.c.u_l_id).select()

    def tearDown(self):
        Base.metadata.drop_all(self.db)
        self.session.commit()
        self.session.close()
        self.db.close()


class TestSqlaQueryCore(DBTestCase):
    def assertEqualColumns(self, first, second):
        first_columns = set(c.name for c in first.columns)
        second_columns = set(c.name for c in second.columns)
        self.assertEquals(first_columns, second_columns)

    def test_no_filters(self):
        sel = self.location.select()
        result = qsqla.query(sel, [])
        self.assertEqualColumns(sel, result)
        self.assertIsNone(result._whereclause)

    def test_single_filter(self):
        result = qsqla.query(self.joined_select, [{"name": "l_id", "op": "eq", "val": 1}])
        self.assertEqualColumns(self.joined_select, result)

    def test_multiple_filters(self):
        filters = [{"name": "l_id", "op": "eq", "val": 1},
                   {"name": "u_id", "op": "eq", "val": 1} ]

        result = qsqla.query(self.joined_select, filters)
        self.assertEqualColumns(self.joined_select, result)

    def test_limit(self):
        limit = 2
        query = qsqla.query(self.user.select(), [], limit=limit)
        rows = self.db.execute(query)
        self.assertEquals(len(list(rows)), limit)

    def test_upper_bound_limit(self):
        query = qsqla.query(self.user.select(), [], limit=1, upper_bound_limit=2)
        rows = self.db.execute(query)
        self.assertEquals(len(list(rows)), 1)

        query = qsqla.query(self.user.select(), [], limit=500, upper_bound_limit=2)
        rows = self.db.execute(query)
        self.assertEquals(len(list(rows)), 2)

    def test_offset(self):
        query = qsqla.query(self.user.select(), [], offset=2)
        rows = self.db.execute(query)
        self.assertEquals(len(list(rows)), 1)

    def test_order_with_default_ascending(self):
        query = qsqla.query(self.user.select(), [], order="u_id")
        rows = self.db.execute(query)
        self.assertEquals([row.u_id for row in rows], [1, 2, 3])

    def test_order_with_forced_ascending(self):
        query = qsqla.query(self.user.select(), [], order="u_id", asc=True)
        rows = self.db.execute(query)
        self.assertEquals([row.u_id for row in rows], [1, 2, 3])

    def test_order_with_forced_descending(self):
        query = qsqla.query(self.user.select(), [], order="u_id", asc=False)
        rows = self.db.execute(query)
        self.assertEquals([row.u_id for row in rows], [3, 2, 1])


class TestSqlaQueryORM(DBTestCase):

    def test_limit(self):
        query = qsqla.query(User, [], limit=2)
        query.session = self.session
        rows = query.all()
        self.assertEquals(len(list(rows)), 2)

    def test_offset(self):
        query = qsqla.query(User, [], offset=2)
        query.session = self.session
        rows = query.all()
        self.assertEquals(len(list(rows)), 1)

    def test_order_with_default_ascending(self):
        query = qsqla.query(User, [], order="u_id")
        query.session = self.session
        rows = query.all()
        self.assertEquals([row.u_id for row in rows], [1, 2, 3])

    def test_order_with_forced_ascending(self):
        query = qsqla.query(User, [], order="u_id", asc=True)
        query.session = self.session
        rows = query.all()
        self.assertEquals([row.u_id for row in rows], [1, 2, 3])

    def test_order_with_forced_descending(self):
        query = qsqla.query(self.user.select(), [], order="u_id", asc=False)
        rows = self.db.execute(query)
        self.assertEquals([row.u_id for row in rows], [3, 2, 1])


class TestOperators(DBTestCase):
    def perform_assertion(self, filter, expected_names):
        # test core
        selectable = qsqla.query(self.joined_select, [filter])
        rows = self.db.execute(selectable)
        self.assertEqual([dict(r)['u_name'] for r in rows], expected_names)

        if filter['name'].startswith('u_'):
            # test ORM
            q = qsqla.query(User, [filter])
            q.session = self.session
            self.assertEqual([row.u_name for row in q.all()], expected_names)

    def test_is_null(self):
        self.perform_assertion({"name": "l_id", "op": "is_null"}, [])
        self.perform_assertion({"name": "u_id", "op": "is_null"}, [])

    def test_is_not_null(self):
        self.perform_assertion({"name": "l_id", "op": "is_not_null"},
                          ['Micha', 'Oli', 'Tom'])
        self.perform_assertion({"name": "u_id", "op": "is_not_null"},
                               ['Micha', 'Oli', 'Tom'])

    def test_equals(self):
        self.perform_assertion({"name": "l_id", "op": "eq", "val": "1"},
                          ['Micha', 'Oli'])
        self.perform_assertion({"name": "u_name", "op": "eq", "val": "Oli"},
                          ['Oli'])
        self.perform_assertion({"name": "u_name", "op": "eq", "val": "Hannes"},
                          [])

    def test_ignore_case_equals(self):
        self.perform_assertion({"name": "u_name", "op": "ieq", "val": "Oli"},
                               ['Oli'])
        self.perform_assertion({"name": "u_name", "op": "ieq", "val": "oli"},
                               ['Oli'])
        self.perform_assertion({"name": "u_name", "op": "ieq", "val": "Hannes"},
                               [])

    def test_not_equals(self):
        self.perform_assertion({"name": "l_id", "op": "ne", "val": "1"},
                          ['Tom'])
        self.perform_assertion({"name": "u_name", "op": "ne", "val": "Oli"},
                          ['Micha', 'Tom'])
        self.perform_assertion({"name": "u_name", "op": "ne", "val": "Hannes"},
                          ['Micha', 'Oli', 'Tom'])

    def test_greater_than(self):
        self.perform_assertion({"name": "u_id", "op": "gt", "val": "1"},
                          ['Oli', 'Tom'])
        self.perform_assertion({"name": "u_id", "op": "gt", "val": "3"},
                          [])

    def test_greater_than_equals(self):
        self.perform_assertion({"name": "u_id", "op": "gte", "val": "1"},
                          ['Micha', 'Oli', 'Tom'])

    def test_less_than(self):
        self.perform_assertion({"name": "u_id", "op": "lt", "val": "1"},
                          [])
        self.perform_assertion({"name": "u_id", "op": "lt", "val": "2"},
                          ['Micha'])

    def test_less_than_equals(self):
        self.perform_assertion({"name": "u_id", "op": "lte", "val": "2"},
                          ['Micha', 'Oli'])

    def test_like(self):
        self.perform_assertion({"name": "l_name", "op": "like", "val": "%gart%"},
                          ['Tom'])
        self.perform_assertion({"name": "u_name", "op": "like", "val": "%om%"},
            ['Tom'])


    # Fails if run with sqlite, see issue https://www.sqlite.org/pragma.html#pragma_case_sensitive_like
    def test_like_is_case_sensitive(self):
        self.perform_assertion(
            {"name": "l_name", "op": "like", "val": "%gaRT%"},
            [])
        self.perform_assertion(
            {"name": "u_name", "op": "like", "val": "%OM%"},
            [])

    def test_not_like(self):
        self.perform_assertion({"name": "l_name", "op": "not_like", "val": "%gart%"},
                          ['Micha', 'Oli'])
        self.perform_assertion(
            {"name": "u_name", "op": "not_like", "val": "%om%"},
            ['Micha', 'Oli'])

    def test_ilike(self):
        self.perform_assertion(
            {"name": "l_name", "op": "ilike", "val": "%gaRT%"},
            ['Tom'])
        self.perform_assertion(
            {"name": "u_name", "op": "ilike", "val": "%oM%"},
            ['Tom'])

    def test_not_ilike(self):
        self.perform_assertion({"name": "l_name", "op": "not_ilike", "val": "%gaRT%"},
                          ['Micha', 'Oli'])
        self.perform_assertion(
            {"name": "u_name", "op": "not_ilike", "val": "%oM%"},
            ['Micha', 'Oli'])

    def test_in_(self):
        self.perform_assertion({"name": "u_id", "op": "in", "val": "1,3"},
                          ['Micha', 'Tom'])

    def test_not_in(self):
        self.perform_assertion({"name": "u_id", "op": "not_in", "val": "1,3"},
                          ['Oli'])

    def test_operation_on_typedecorated_type(self):
        val = (datetime.now() - timedelta(hours=5)).isoformat()
        self.perform_assertion({"name": "u_date", "op": "ne", "val": val},
                               ['Micha', 'Oli', 'Tom'])

    def test_greater_than_typedecorated_datetime(self):
        datestring = (self.now - timedelta(minutes=1)).isoformat()
        self.perform_assertion({"name": "u_date", "op": "gt", "val": datestring},
                          ['Micha', 'Oli', 'Tom'])
        datestring = (self.now + timedelta(minutes=1)).isoformat()
        self.perform_assertion(
            {"name": "l_date", "op": "gt", "val": datestring},
            [])

    def test_with_eq_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__eq=Hooch"}], order="u_id", asc=True)
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_ne_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_id__ne=3"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_ieq_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__ieq=beethoven"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha'])

    def test_with_gt_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_id__gt=2"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Tom'])

    def test_with_lt_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_id__lt=2"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha'])

    def test_with_gte_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_id__gte=2"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli', 'Tom'])

    def test_with_lte_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_id__lte=2"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_like_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__like=%eeth%"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha'])

    def test_with_not_like_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__not_like=%issy"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_ilike_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__ilike=%eeth%"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha'])

    def test_with_not_ilike_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__not_ilike=%issy"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_in_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__in=Beethoven,Sissy"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Tom'])

    def test_with_not_in_many_relation(self):
        q = qsqla.query(User, [{"name": "pets", "op": "with", "val": "p_name__not_in=Hooch,Sissy"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha'])

    def test_with_eq_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__eq=Stuttgart"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Tom'])

    def test_with_ne_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_id__ne=1"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Tom'])

    def test_with_ieq_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__ieq=karlsruhe"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_gt_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_id__gt=1"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Tom'])

    def test_with_lt_Scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_id__lt=2"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_gte_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_id__gte=2"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Tom'])

    def test_with_lte_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_id__lte=1"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_like_location_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__like=%uttg%"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Tom'])

    def test_with_not_like_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__not_like=%uttga%"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_ilike_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__ilike=%ArlSr%"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_not_ilike_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__not_ilike=%uttgart"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])

    def test_with_in_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__in=Karlsruhe,Stuttgart"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli', 'Tom'])

    def test_with_not_in_scalar_relation(self):
        q = qsqla.query(User, [{"name": "location", "op": "with", "val": "l_name__not_in=Stuttgart"}])
        q.session = self.session
        self.assertEqual([row.u_name for row in q.all()], ['Micha', 'Oli'])


qsqla/query.py
"""
qSQLA Query Syntax
==================

qSQLA is a Query Syntax to filter flat records derived from SQLAlchemy selectable objects.
Each field can be queried with a number of different operators.

The filters are provided in the query string of a ``HTTP GET`` request. The operator is added with a double underscore
to the field name. Unary Operators do not need to specify a value.

.. code::

    GET http://host/vsi/log/deliveries?delivery_id__eq=55&delivery_date__gt=2016-01-01T01:00:00

    Filter:
        delivery_id = 55
        delivery_date > 2016-01-01T01:00:00

    Response:
        [{u'delivery_id': 55,
          u'delivery_category': u'Locations',
          u'delivery_date': u'2016-06-14T06:46:02.296028+00:00',
          u'id': 42,
          u'create_date': u'2016-06-14T06:46:02.296028+00:00',
          u'error_info': None,
          u'row_count': 1,
          u'state': 2,
          u'update_date': u'2016-06-14T06:46:02.296028+00:00'}]

Supported Operators are:

Unary operators:

- ``is_null`` for all fields
- ``is_not_null`` for all fields
- ``is_true`` for Boolean fields
- ``is_false`` for Boolean fields

Binary operators:

- ``eq`` for Integer, String, Date and DateTime fields
- ``ne`` for Integer, String, Date and DateTime fields
- ``ieq`` case insentitive equal for String fields
- ``gt`` for Integer, Date and DateTime fields
- ``lt`` for Integer, Date and DateTime fields
- ``gte`` for Integer, Date and DateTime fields
- ``lte`` for Integer, Date and DateTime fields
- ``like`` for String fields. Case-insensitivity is dependant on the database implementation of LIKE.
- ``not_like`` for String fields. Case-insensitivity is dependant on the database implementation of NOT LIKE.
- ``ilike`` always case-insensitive like for String fields
- ``not_ilike`` always case-insensitive not like for String fields
- ``in`` for Integer, String fields. The values are provided as a comma separated list.
- ``not_in`` for Integer, String fields. The values are provided as a comma separated list.
- ``with`` for relationships. Combine any other binary operator with an additional `__` on any relationship. Can only be used on ORM queries.

Supported Types:

- ``sqlalchemy.types.Integer``
- ``sqlalchemy.types.Boolean``
- ``sqlalchemy.types.Date``
- ``sqlalchemy.types.DateTime``
- ``sqlalchemy.types.String``

In addition to the filters one can provide

- ``_limit`` Limit the query to a number of records.
- ``_offset`` Add an offset to the query.
- ``_order``  The order field.
- ``_desc`` If provided sort in descending order, else in ascending.

"""
import functools

import dateutil.parser
import sqlalchemy
from sqlalchemy.sql.selectable import Selectable


def requires_types(*types):
    def dec(f):
        @functools.wraps(f)
        def wrapper(arg1, arg2=None):
            arg_basetype = getattr(arg1.type, 'impl', arg1.type)
            if not any([isinstance(arg_basetype, t) for t in types]):
                raise TypeError("Cannot apply filter to field {}".format(arg1.name))
            return f(arg1, arg2)
        return wrapper
    return dec


def requires_mapped_attribute(f):
    @functools.wraps(f)
    def wrapper(arg1, arg2=None):
        try:
            getattr(arg1.property, 'mapper')
        except:
            raise TypeError("{} is not a mapped attribute".format(arg1))
        return f(arg1, arg2)
    return wrapper


def convert_type(type_, value):
    cls = type_.__class__
    basetype = getattr(cls, 'impl', cls)
    if issubclass(basetype, sqlalchemy.types.Integer):
        return int(value)
    elif issubclass(basetype, sqlalchemy.types.String):
        return value
    elif issubclass(basetype, sqlalchemy.types.DateTime):
        return dateutil.parser.parse(value)


def get_subquery(arg1, arg2):
    if "=" not in arg2:
        raise KeyError("`any` expects a second operator added with `__`")
    subquery = split_operator(arg2)
    subquery_column = getattr(arg1.property.mapper.class_, subquery[0])
    subquery_op_val = subquery[1].rsplit("=", 1)
    return subquery_column, subquery_op_val


def convert_generic(f):
    @functools.wraps(f)
    def wrapper(arg1, arg2=None):
        return f(arg1, convert_type(arg1.type, arg2))
    return wrapper

def convert_list(f):
    @functools.wraps(f)
    def wrapper(arg1, arg2=None):
        vals = [convert_type(arg1.type, arg.strip()) for arg in arg2.split(",")]
        return f(arg1, vals)
    return wrapper


def is_null(arg1, arg2=None):
    return arg1 == None  # NOQA


def is_not_null(arg1, arg2=None):
    return arg1 != None  # NOQA


@requires_types(sqlalchemy.types.Boolean)
def is_true(arg1, arg2=None):
    return arg1 == True  # NOQA


@requires_types(sqlalchemy.types.Boolean)
def is_false(arg1, arg2=None):
    return arg1 == False  # NOQA


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.String,
                sqlalchemy.types.DateTime)
@convert_generic
def equals(arg1, arg2):
    return arg1 == arg2


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.String,
                sqlalchemy.types.DateTime)
@convert_generic
def not_equals(arg1, arg2):
    return arg1 != arg2


@requires_types(sqlalchemy.types.String)
@convert_generic
def ignore_case_equals(arg1, arg2):
    return sqlalchemy.func.lower(arg1) == arg2.lower()


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.DateTime)
@convert_generic
def greater_than(arg1, arg2):
    return arg1 > arg2


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.DateTime)
@convert_generic
def greater_than_equals(arg1, arg2):
    return arg1 >= arg2


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.DateTime)
@convert_generic
def less_than(arg1, arg2):
    return arg1 < arg2


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.DateTime)
@convert_generic
def less_than_equals(arg1, arg2):
    return arg1 <= arg2


@requires_types(sqlalchemy.types.String)
@convert_generic
def like(arg1, arg2):
    return arg1.like(arg2)


@requires_types(sqlalchemy.types.String)
@convert_generic
def not_like(arg1, arg2):
    return ~arg1.like(arg2)


@requires_types(sqlalchemy.types.String)
@convert_generic
def ilike(arg1, arg2):
    return arg1.ilike(arg2)


@requires_types(sqlalchemy.types.String)
@convert_generic
def not_ilike(arg1, arg2):
    return ~arg1.ilike(arg2)


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.String)
@convert_list
def in_(arg1, arg2):
    return arg1.in_(arg2)


@requires_types(sqlalchemy.types.Integer, sqlalchemy.types.String)
@convert_list
def not_in(arg1, arg2):
    return ~arg1.in_(arg2)


@requires_mapped_attribute
def with_(arg1, arg2):
    subquery_column, subquery_op_val = get_subquery(arg1, arg2)
    if arg1.property.uselist:
        restriction = arg1.any(OPERATORS[subquery_op_val[0]](subquery_column, subquery_op_val[1]))
    else:
        restriction = arg1.has(OPERATORS[subquery_op_val[0]](subquery_column, subquery_op_val[1]))
    return restriction


UNARY_OPERATORS = ['is_null', 'is_not_null', 'is_true', 'is_false']


OPERATORS = {
    # Unary operators.
    'is_null': is_null,
    'is_not_null': is_not_null,
    'is_true': is_true,
    'is_false': is_false,
    # Binary operators.
    'eq': equals,
    'ne': not_equals,
    'ieq': ignore_case_equals,
    'gt': greater_than,
    'lt': less_than,
    'gte': greater_than_equals,
    'lte': less_than_equals,
    'like': like,
    'not_like': not_like,
    'ilike': ilike,
    'not_ilike': not_ilike,
    'in': in_,
    'not_in': not_in,
    'with': with_
}


def split_operator(param):
    query = param.rsplit('__', 1)
    if len(query) == 1:
        name = query[0]
        operator = 'eq'
    else:
        name = query[0]
        operator = query[1].lower() if (len(query[1]) == 2) else query[1]
    if name == '':
        raise ValueError("No valid parameter provided")
    return (name, operator)


def build_filters(query):
    """ build filter dictionary from a query dict"""
    filters = []
    for key, val in query.items():
        if len(key.rsplit('__', 3)) == 4:
            keys = key.rsplit('__', 3)
            name = keys[0]
            operator = keys[1]
            val = "{}__{}={}".format(keys[2], keys[3], val)
        else:
            name, operator = split_operator(key)
        filters.append({"name": name, "op": operator, "val": val})
    return filters


def get_column(s, name):
    for col in s.columns:
        if col.name.lower() == name.lower():
            return col
    raise KeyError("column {} not found".format(name))


def query(selectable_or_model, filters, limit=None, offset=None, order=None,
          asc=True, upper_bound_limit=10000):
    """
    Main entry point for applying filters and pagination controls.

    :param selectable_or_model: an SQLAlchemy Core Selectable or ORM Model
    :param filters: a list of filters produced by build_filters
    :param limit: int. the limit.
    :param offset: int. the offset.
    :param order: string. The name of the field to order by.
    :param asc: bool. Ascending (default) or descending order.
    :param upper_bound_limit: int. An absolute upper bound limit to use. Disabled if set to None.

    :raises KeyError: if key is not available in query
    :raises ValueError: if value cannot be converted to Column Type
    :raises TypeError: if filter is not available for SQLAlchemy Column Type

    :return: an SQLAlchemy Core Selectable or ORM Query object.
    """
    use_core = isinstance(selectable_or_model, Selectable)
    func = core_query if use_core else orm_query
    filtered = func(selectable_or_model, filters)

    if order:
        if use_core:
            order_col = get_column(selectable_or_model, order)
        else:
            order_col = getattr(selectable_or_model, order)
        if not asc:
            order_col = order_col.desc()
        filtered = filtered.order_by(order_col)

    if limit or upper_bound_limit:
        if limit is None:
            limit = upper_bound_limit
        else:
            if upper_bound_limit:
                limit = min(int(limit), upper_bound_limit)
            else:
                limit = int(limit)
        filtered = filtered.limit(limit)

    if offset:
        filtered = filtered.offset(offset)

    return filtered


def core_query(selectable, filters):
    """Add filters to an sqlalchemy selectable

    :param selectable: the select statements
    :param filters: a list of filters produced by build_filters

    :raises KeyError: if key is not available in query
    :raises ValueError: if value cannot be converted to Column Type
    :raises TypeError: if filter is not available for SQLAlchemy Column Type

    :return: a selectable with the filters applied
    """
    restrictions = []

    alias = selectable.alias("query")

    for f in filters:
        col = get_column(alias, f["name"])
        if f["op"] in UNARY_OPERATORS:
            restrictions.append(OPERATORS[f["op"]](col))
        else:
            restrictions.append(OPERATORS[f["op"]](col, f["val"]))

    if restrictions:
        sel = sqlalchemy.select([alias], whereclause=sqlalchemy.and_(*restrictions))
    else:
        sel = sqlalchemy.select([alias])
    return sel


def orm_query(model, filters):
    """ Add filters to an sqlalchemy ORM query
    :param model: an SQLAlchemy Model
    :param filters: a list of filters produced by build_filters

    :return: a SQLAlchemy ORM Query with the filters applied
    """
    query = sqlalchemy.orm.Query(model)
    restrictions = []
    for f in filters:
        col = getattr(model, f['name'])
        if f["op"] in UNARY_OPERATORS:
            restrictions.append(OPERATORS[f["op"]](col))
        else:
            restrictions.append(OPERATORS[f["op"]](col, f["val"]))
    query = query.filter(*restrictions)
    return query
